### PR TITLE
refactor!: decouple state models from conversion logic via codec layer

### LINKED ISSUE #1079: Break import cycles between subpackages to enable module-boundary DAG enforcement
## Context

`tools/check_module_boundaries.py` is an AST-based CI guard that forbids upward/peer imports between hassette subpackages (exempting `TYPE_CHECKING`-guarded imports). The full layer DAG and cycle-freedom from #633 cannot be enforced while the runtime subpackage graph still contains cycles. This issue is the refactoring that removes those cycles; #633 is the enforcement it unblocks.

> **Status update (2026-06-22):** This issue was filed against an older tree. Since then **#1097** (moved `await_guard` to `utils/`, extracted `hassette.schemas`, straightened `utils ↔ events`; landed `api-no-core`, `utils-no-events`, `web-no-core`) and **#1103** (moved `InvokeHandler`/`ExecuteJob` to a root-leaf `hassette/commands.py`; landed `bus-no-core`) resolved most of the originally-enumerated cycles. Both `check_module_boundaries.py` and `check_lazy_imports.py` currently pass clean. The remaining work is narrower than the original list below — see "Remaining cycles."

## Remaining cycles

Four runtime cycles remain. Two are the hard service-layer edges every prior artifact deferred; two are small data-model cycles still held open by annotated lazy imports.

| Cycle | Back-edge (wrong direction) | Resolution |
|---|---|---|
| `core ↔ scheduler` | `scheduler/scheduler.py:74` imports `SchedulerService` from `core` (live top-level import) | **Protocol inversion** — `SchedulerServiceProtocol` (7 members) in L0 `types`; `Scheduler` depends on the protocol, `SchedulerService` stays in `core` and satisfies it structurally |
| `core ↔ state_manager` | `state_manager/state_manager.py:10` imports `StateProxy` from `core` (live top-level import) | **Protocol inversion** — read-only `StateReader` (4 members: `get_state`, `num_domain_states`, `yield_domain_states`, `__contains__`) in L0 `types` |
| `resources ↔ task_bucket` | `resources/base.py:145` lazy `TaskBucket` import | Structural — inject `TaskBucket` (or a tiny protocol/factory) instead of importing inside `Resource.__init__` |
| `conversion ↔ models` (#892) | `conversion/annotation_converter.py:38,40`, `conversion/state_registry.py:89` lazy `BaseState`/`TYPE_MATCHER` imports | Structural — relocate the registry seam so `models.states.base` no longer imports `conversion` at runtime (or demote `BaseState` to `TYPE_CHECKING` where annotation-only and move the runtime `issubclass`/registry call below the boundary) |

### Why protocol inversion (not relocation) for the two core cycles

`SchedulerService` and `StateProxy` import **nothing** from their consumer packages — each back-edge is a single import line. But each imports 3–4 `core` sibling services (`database_service`, `sync_executor_service`, `websocket_service`, …) that genuinely cannot move down. Relocating the classes into `scheduler/`/`state_manager/` (L5) would convert two clean cycles into several **new** L5→core upward violations. Protocol inversion avoids that, follows the existing `TriggerProtocol` precedent (already in L0 `types`), and improves testability — `DomainStates` today leaks `StateProxy` reconnection internals into the public state API. No ADR — the decision is recorded in the design doc.

## Plan (two waves, simplest-first)

**Wave 1 — small cycles (no design decision, parallelizable):**
- [ ] `resources ↔ task_bucket`: remove the `resources/base.py:145` lazy import; delete the `# lazy-import:` annotation.
- [ ] `conversion ↔ models` (#892): remove the three `conversion` lazy imports; delete the annotations.

**Wave 2 — core cycles via protocol inversion:**
- [ ] `core ↔ scheduler`: define `SchedulerServiceProtocol` in `types`; retype `scheduler/scheduler.py` against it; remove the runtime `from hassette.core.scheduler_service import SchedulerService`; add a `scheduler-no-core` rule.
- [ ] `core ↔ state_manager`: define `StateReader` in `types`; retype `DomainStates`/`StateManager`; remove the runtime `from hassette.core.state_proxy import StateProxy`; add a `state_manager-no-core` rule.

**Enforcement (#633, strictly last):**
- [ ] Expand `check_module_boundaries.py` `RULES` to the full L0–L9 layer map and add a hand-written graph cycle detector. Tracked under #633; this issue is its blocker.

## Acceptance Criteria

- [ ] The four remaining runtime cycles are broken; the subpackage runtime import graph is acyclic.
- [ ] The cycle-related `# lazy-import:` annotations are removed (`resources/base.py`, the three `conversion` sites). Note: `__main__.py:9` (CLI startup deferral) and `conversion/validation.py:77` (load-order note) are **not** cycles and stay.
- [ ] Each core-cycle fix adds a self-proving boundary rule (`scheduler-no-core`, `state_manager-no-core`) so the break can't re-accrete.
- [ ] `check_module_boundaries.py` and `check_lazy_imports.py` pass; pyright clean; test suite green.

## Notes

Each cycle is independently resolvable and could be split into per-cycle sub-issues (Wave 1 items parallelize; the two core-cycle PRs both touch `core.py` wiring, so coordinate or serialize those edits). Research brief: `design/research/2026-06-22-break-import-cycles/research.md`. Related: #633 (enforcement this unblocks), #892 (`conversion ↔ models`).

### PR DESCRIPTION
**Pull request by NodeJSmith**
State models (`BaseState` subclasses) carried value coercion logic inside a Pydantic `@model_validator` — converting `"on"` to `True`, parsing numeric strings, normalizing `"unknown"`/`"unavailable"`. This forced a runtime import cycle between `models` and `conversion`, since the validator reached for `TYPE_REGISTRY.convert()` at validation time. That cycle was the last one blocking DAG enforcement (#1079), and resisted the mechanical fixes that cleared the other three cycles in #1120.

This PR moves all value coercion out of the models and into a codec layer owned by `conversion/`, making models "dumb" Pydantic data containers. See `design/adrs/0003-dumb-state-models-codec-owned-conversion.md` for the full rationale and `design/specs/084-dumb-state-models-codec-conversion/design.md` for the implementation approach.

### Codec layer

- `StateRegistry.coerce_and_construct()` is the new entry point for codec-owned value coercion — preprocessing (domain extraction, unknown/unavailable normalization) followed by `TYPE_REGISTRY.convert()` for the model's `value_type`
- `state_manager` and dependency extractors (`StateNew`, `StateOld`, etc.) route through the codec instead of relying on model self-validation
- `convert_state_dict_to_model()` applies the same preprocessing order as the old validator so behavior is identical — confirmed by characterization tests

### Import cycle fix

- New `models/states/catalog.py` module acts as a leaf: `BaseState.__init_subclass__` registers into the catalog (a plain dict), and `conversion/` reads from it — no back-edge from models → conversion
- `STATE_REGISTRY` and `TYPE_REGISTRY` singletons relocated from `models/` to `conversion/`, with re-exports for backwards compatibility
- All three `# lazy-import:` markers in `conversion/` replaced with normal top-level imports
- `check_module_boundaries.py` updated to enforce the new DAG edge rules

### Test coverage

- Characterization tests pin codec behavior across every `value_type` family: bool, string, numeric, datetime, time, and unknown/unavailable normalization
- `DomainStates` integration tests verify the full `state_manager → codec → model` path produces typed values
- State catalog tests verify `__init_subclass__` registration depth and domain resolution

### Documentation

- `states/conversion.md` and `states/custom-states.md` updated to describe codec-owned conversion and correct `value_type` descriptions for `BoolBaseState` and `NumericBaseState`

BREAKING CHANGE: `STATE_REGISTRY` and `TYPE_REGISTRY` are now owned by `hassette.conversion` instead of `hassette.models`. Existing imports from `hassette.models` continue to work via re-exports but are deprecated — update imports to `from hassette.conversion import STATE_REGISTRY, TYPE_REGISTRY`.

Closes #1079

### CODE REVIEW

File: src/hassette/conversion/state_registry.py [hunk 1]
```diff
@@ -175,55 +151,50 @@ def conversion_with_error_handling(
...
+    state = prepared.get("state")
+    if state == "unknown":
+        prepared["is_unknown"] = True
+        prepared["state"] = state = None
+    elif state == "unavailable":
+        prepared["is_unavailable"] = True
+        prepared["state"] = state = None
+
+    with suppress(UnableToConvertValueError):
+        prepared["state"] = TYPE_REGISTRY.convert(state, model.value_type)
```
[the remark this answers is not in the document]
Comment (NodeJSmith) [reply src/hassette/conversion/state_registry.py]: Fixed — guarded state normalization with `"state" in prepared` so a missing state field is no longer synthesized during conversion.

File: tests/unit/conversion/test_codec_characterization.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
+"""Characterization pin for state-conversion typed output.
+
+This test captures the exact typed values produced by the current model_validate
+path across every value_type family. It must stay green before and after the
+refactor — if any assertion fails post-refactor, the codec produced a different
+typed value than the pre-change path.
+
+Write golden values from observation, not from calling the same code path twice.
+"""
```
[the remark this answers is not in the document]
Comment (NodeJSmith) [reply tests/unit/conversion/test_codec_characterization.py:9]: Fixed — updated the module docstring to accurately describe which entry points each test class covers, and replaced string name comparisons with direct type checks (`type(state) is LightState`).

**Comment by NodeJSmith** [comment]
Addressed the BaseState fallback gap and broader custom conversion test coverage in ad475c48:

- Added `test_unregistered_domain_falls_back_to_base_state` to `TestTryConvertStateEntryPoint` (the specific finding)
- Added type assertion on `test_unknown_light_via_registry`
- New `test_custom_conversion.py` with 22 tests covering custom state attributes, value coercion through built-in base classes, custom StrEnum value_type with registered converters, `@register_type_converter_fn` decorator, and `convert_state_dict_to_model` direct calls
